yellow-paper/docs/contracts/main.md

```diff
@@ -0,0 +1,319 @@
+While this can be addressed by having a queue of messages and let the sequencer specify the order, this require extra logic and might be difficult to price correctly. To keep this out of protocol, we simply allow the user to attach a fee to the message (see `fee` in `L1ToL2Msg` above). This way, the user can incentivize the sequencer to pick up the message faster.
+
+Since it is possible to land in a case where the sequencer will never pick up the message (e.g., if it is underpriced), the sender must be able to cancel the message. To ensure that this cancellation cannot happen under the feet of the sequencer we use a `deadline`, only after the deadline can it be cancelled. 
+
+The contract that sent the message must itself decide how it is to handle the cancellation. It could be that the contract simply ignores the message, or it could be that it refunds the user. This is up to the contract to decide. 
+
+:::info Error handling
+While we have ensured that the message either arrives to the L2 outbox or is cancelled, we have not ensured that the message is consumed by the L2 contract. This is up to the L2 contract to handle. If the L2 contract does not handle the message, it will be stuck in the outbox forever. Similarly, it is up to the L1 contract to handle the cancellation. If the L1 contract does not handle the cancellation, the user might have a message that is pending forever. Error handling is entirely on the contract developer.
+:::
+
+##### L2 Inbox
+While the L2 inbox is not a real contract, it is a logical contract that apply mutations to the data similar to the L1 inbox to ensure that the sender cannot fake his position. This logic is handled by the kernel and rollup circuits.
+
+As for the L1 variant, we must populate the some fields:
+- `L2Actor.actor`: The sender of the message (the caller) [also in L1 inbox]
+- `L2Actor.version`: The version of the L2 chain sending the message [also in L1 inbox]
+- `L1Actor.actor` The recipient of the message (the portal)
+- `L1Actor.chainId` The chainId of the L1 chain receiving the message
+
+In practice, this is done at the kernel layer of the L2, and the message hash is a public output of the circuit that is inserted into the L1 outbox for later consumption.
+
+:::warning Comment for discussion
+Note that while we are letting the inbox populate more values that what we did for the L1 inbox. This is more an opinionated decision than a purely technical one. 
+
+We could let the contract itself populated the `L1Actor` like we did for L1, but we decided to let the kernel do it instead, since access control can be quite tedious to get right in private execution. By having the `portal` contract that is specified at the time of deployment, we can insert this value and ensure that it is controlled by the contract.
+If we have a better alternative for access control (e.g., slow updates), this could be changed to be more similar to the L1 inbox, which gives better flexibility.
```

Comment:
```suggestion
If we have a better alternative for access control, this could be changed to be more similar to the L1 inbox, which gives better flexibility.
```